contracts/halo-pool/src/contract.rs

```diff
@@ -139,114 +196,320 @@ pub fn execute_add_reward_balance(
+    let current_time = env.block.time.seconds();
+    // Get pool infos
+    let mut pool_infos = POOL_INFOS.load(deps.storage)?;
+    // Get current pool index
+    let current_phase_index = pool_infos.current_phase_index;
+    // Get current pool info in pool infos
+    let pool_info = pool_infos.pool_infos[current_phase_index as usize].clone();
     // get staker info
     let mut staker_info = STAKERS_INFO
         .may_load(deps.storage, info.sender.clone())?
-        .unwrap_or(StakerRewardAssetInfo {
+        .unwrap_or(StakerInfoResponse {
             amount: Uint128::zero(),
-            reward_debt: Uint128::zero(),
+            reward_debt: vec![Uint128::zero(); current_phase_index as usize + 1],
+            joined_phase: current_phase_index,
         });
+
+    // Not allow depositing when current time is greater than end time of the phase
+    // and less than start time of the phase
+    if current_time > pool_info.end_time || current_time < pool_info.start_time {
+        return Err(ContractError::Std(StdError::generic_err(
+            "Current time is not in the range of the phase",
+        )));
+    }
+
     // Check pool limit per user
     if let Some(pool_limit_per_user) = pool_info.pool_limit_per_user {
         if staker_info.amount + amount > pool_limit_per_user {
             return Err(ContractError::Std(StdError::generic_err(
-                "Unauthorized: Deposit amount exceeds pool limit per user",
+                "Deposit amount exceeds pool limit per user",
             )));
         }
     }
 
-    // Get last reward time
-    let last_reward_time = LAST_REWARD_TIME.load(deps.storage)?;
-
-    // Get accrued token per share
-    let accrued_token_per_share = ACCRUED_TOKEN_PER_SHARE.load(deps.storage)?;
-
+    // Init reward amount
+    let mut reward_amount = Uint128::zero();
+    // Init new accrued token per share
+    let mut new_accrued_token_per_share;
+    // Init new last reward time
+    let mut new_last_reward_time;
+    // Init response
     let mut res = Response::new();
 
+    // If staker has joined previous phases, loops all pool infos to get reward per second from current pool index to staker joined phases
+    if staker_info.joined_phase < current_phase_index {
+        for i in staker_info.joined_phase..current_phase_index {
+            // Get last reward time
+            let last_reward_time = pool_infos.last_reward_time[i as usize];
+            // Get accrued token per share
+            let accrued_token_per_share = pool_infos.accrued_token_per_share[i as usize];
+            // Get pool info from pool infos
+            let pool_info = POOL_INFOS.load(deps.storage)?.pool_infos[i as usize].clone();
+            // Get staked token balance from pool contract previous phases
+            let staked_token_supply = pool_info.total_staked_at_end_time;
+
+            // update pool
+            (new_accrued_token_per_share, new_last_reward_time) = update_pool(
```

Comment:
hàm này có lẽ ko nên để tên là update pool vì nó ko thay đổi data j của pool cả. (get_debt_value???)

Reply:
Changed to: get_new_reward_ratio_and_time